File: truth_set/python/structure/organisation/long-sub-routine/ReportService.py
import os
from datetime import datetime

class ReportService:
    def __init__(self):
        pass
        
    def generate_monthly_report(self, users, transactions):
        active_users = [user for user in users if user['is_active']]
        print(f"Active Users Count: {len(active_users)}")

        total_amount = 0
        for tx in transactions:
            total_amount += tx['amount']
        print(f"Total Transactions Amount: ${total_amount}")

        report_content = f"""Monthly Report
================
Active Users: {len(active_users)}
Total Transactions: ${total_amount}
"""

        reports_dir = os.path.join(os.path.dirname(__file__), 'reports')
        if not os.path.exists(reports_dir):
            os.makedirs(reports_dir)
            
        file_path = os.path.join(reports_dir, f"monthly_report_{datetime.now().month}.txt")
        with open(file_path, 'w') as f:
            f.write(report_content)
            
        print(f"Report saved to {file_path}")

        self.send_notification(f"Monthly report generated: {file_path}")

    def send_notification(self, message):
        print(f"Notification: {message}")


# Test data
users = [
    {'id': 1, 'name': 'Alice', 'is_active': True},
    {'id': 2, 'name': 'Bob', 'is_active': False}, 
    {'id': 3, 'name': 'Charlie', 'is_active': True}
]

transactions = [
    {'id': 101, 'user_id': 1, 'amount': 250},
    {'id': 102, 'user_id': 3, 'amount': 450},
    {'id': 103, 'user_id': 1, 'amount': 150}
]

# Generate report
report_service = ReportService()
report_service.generate_monthly_report(users, transactions)

File: truth_set/go/structure/organisation/long-sub-routine/ReportService.go
package reportservice

import (
	"fmt"
	"os"
	"path/filepath"
	"time"
)

type User struct {
	ID       int
	Name     string
	IsActive bool
}

type Transaction struct {
	ID     int
	UserID int
	Amount float64
}

type ReportService struct{}

func (rs *ReportService) generateMonthlyReport(users []User, transactions []Transaction) error {
	activeUsers := []User{}
	for _, user := range users {
		if user.IsActive {
			activeUsers = append(activeUsers, user)
		}
	}
	fmt.Printf("Active Users Count: %d\n", len(activeUsers))

	var totalAmount float64
	for _, tx := range transactions {
		totalAmount += tx.Amount
	}
	fmt.Printf("Total Transactions Amount: $%.2f\n", totalAmount)

	reportContent := fmt.Sprintf("Monthly Report\n================\nActive Users: %d\nTotal Transactions: $%.2f\n",
		len(activeUsers), totalAmount)

	reportsDir := filepath.Join(".", "reports")
	if err := os.MkdirAll(reportsDir, 0755); err != nil {
		return fmt.Errorf("failed to create reports directory: %v", err)
	}

	month := time.Now().Month()
	filePath := filepath.Join(reportsDir, fmt.Sprintf("monthly_report_%d.txt", month))
	if err := os.WriteFile(filePath, []byte(reportContent), 0644); err != nil {
		return fmt.Errorf("failed to write report file: %v", err)
	}
	fmt.Printf("Report saved to %s\n", filePath)

	rs.sendNotification(fmt.Sprintf("Monthly report generated: %s", filePath))
	return nil
}

func (rs *ReportService) sendNotification(message string) {
	fmt.Printf("Notification: %s\n", message)
}

func main() {
	users := []User{
		{ID: 1, Name: "Alice", IsActive: true},
		{ID: 2, Name: "Bob", IsActive: false},
		{ID: 3, Name: "Charlie", IsActive: true},
	}

	transactions := []Transaction{
		{ID: 101, UserID: 1, Amount: 250},
		{ID: 102, UserID: 3, Amount: 450},
		{ID: 103, UserID: 1, Amount: 150},
	}

	reportService := &ReportService{}
	reportService.generateMonthlyReport(users, transactions)
}
\ No newline at end of file